src/main/scala/com/microsoft/hyperspace/actions/CreateActionBase.scala

```diff
@@ -72,6 +72,13 @@ private[actions] abstract class CreateActionBase(dataManager: IndexDataManager)
           LogicalPlanFingerprint(
             LogicalPlanFingerprint.Properties(Seq(Signature(signatureProvider.name, s)))))
 
+        /*
+         * Include the project version in the index log entry file.
+         * This may help prevent in mismatching of project version and index
+         * log entry files.
+         */
+        val projectVersion: Option[(String, String)] = Some("projectVersion", BuildInfo.version)
```

Comment:
How about the following?
In `case class IndexLogEntry`, add the following `require`:
```scala
case class IndexLogEntry(
    name: String,
    derivedDataset: CoveringIndex,
    content: Content,
    source: Source,
    properties: Map[String, String])
    extends LogEntry(IndexLogEntry.VERSION) {
  require(properties(IndexLogEntry.HYPERSPACE_VERSION_PROP_KEY) == BuildInfo.version)
  ...
```

In `object IndexLogEntry`, add the following:
```scala
  private val HYPERSPACE_VERSION_PROP_KEY = "hyperspaceVersion"

  def create(
    name: String,
    derivedDataset: CoveringIndex,
    content: Content,
    source: Source,
    properties: Map[String, String]): IndexLogEntry = {
    IndexLogEntry(
      name,
      derivedDataset,
      content,
      source,
      properties + (HYPERSPACE_VERSION_PROP_KEY -> BuildInfo.version))
```

And here, you can just call `IndexLogEntry.create`. This should ensure to always have the property and may help when we support multiple index types.

Reply:
`require(properties(IndexLogEntry.HYPERSPACE_VERSION_PROP_KEY) == BuildInfo.version)` could be too restrictive now since we need to handle the older version without this property.

Reply:
We can introduce this (`require(properties.contains(IndexLogEntry.HYPERSPACE_VERSION_PROP_KEY)`) when we introduce a breaking change in the log entry.

Reply:
I believe what you are proposing is that you want the projectVersion property to be defined inside IndexLogEntry rather than from multiple places outside. Where best scenario is that it is a required property and should be checked inside while creating the case class instance.

However i don't like your approach as we now have two ways to create IndexLogEntry one from create static method and the other using the case class itself. We also can't make case class private as it is needed throughout code.

I propose we make the properties map mutable and add the property projectVersion if it is not present. IndexLogEntry is a dev class so having a mutable property should be fine? Also changing the equals method to compare properties. 

That way we don't introduce the create method and the prroperty becomes default and required in the construction of the instance of IndexLogEntry without breaking backward compatibility.

Let me know your thoughts, if I missed something. If so I can revert to the create approach.

Reply:
seems like having mutable as a public field is worse since it opens up the possibility of modifying the object more?

having `require` (after a breaking change) seems like a reasonable tradeoff to me. (btw, you can't just naively set one now on the deserialize path since it wouldn't be semantically correct)

Reply:
Fair point, even though it is dev code someone may end up breaking it. Changed back to mutable.

I read online just now we can overwrite the case class constructor apply method, so instead of create method i introduced apply and there I added the property.

Also I kept the require statement, which only checks `projectVersion` with respect to project build property only if incoming `properties` contain `projectVersion` in the first place.

Reply:
fair enough, i will change it to immutable.

We might need to maintain a set of combinations of index log file project Version supported to buildVersion, i guess you are correct, will remove require for now.

Reply:
yes. we need to store a min compatible version. Could you help driving this effort as well separately? cc @rapoth

Reply:
Okay, sure!

Reply:
Can you please open an issue and link here when you get a chance? Thank you!